src/analysis/normed_space/mazur_ulam.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2020 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Yury Kudryashov
+-/
+import analysis.normed_space.point_reflection
+import topology.instances.real_vector_space
+
+/-!
+# Mazur-Ulam Theorem
+
+Mazur-Ulam theorem states that an isometric bijection between two normed spaces over `ℝ` is affine.
+Since `mathlib` has no notion of an affine map (yet?), we formalize it in two definitions:
+
+* `isometric.to_real_linear_equiv_of_map_zero` : given `E ≃ᵢ F` sending `0` to `0`,
+  returns `E ≃L[ℝ] F` with the same `to_fun` and `inv_fun`;
+* `isometric.to_real_linear_equiv` : given `f : E ≃ᵢ F`,
+  returns `g : E ≃L[ℝ] F` with `g x = f x - f 0`.
+
+The formalization is based on [Jussi Väisälä, *A Proof of the Mazur-Ulam Theorem*][Vaisala_2003].
+
+## Tags
+
+isometry, affine map, linear map
+-/
+
+variables {E : Type*} [normed_group E] [normed_space ℝ E]
+  {F : Type*} [normed_group F] [normed_space ℝ F]
+
+open set
+
+noncomputable theory
+
+namespace isometric
+
+/-- If an isometric self-homeomorphism of a normed vector space over `ℝ` fixes `x` and `y`,
+then it fixes the midpoint of `[x, y]`. This is a lemma for a more general Mazur-Ulam theorem,
+see below. -/
+lemma midpoint_fixed {x y : E} :
```

Comment:
AFAIR, it was easier for me to have a separate lemma for the part of the proof using only one space. This way I can avoid using some `generalize`; `clear` lines in the main proof.

Reply:
Since this is the interesting part of the proof, where something happens, I would prefer it you could it separate. In a paper proof, it would certainly be singled out at the crucial lemma.

Reply:
Ok, no problem.